### PR TITLE
[TwigBundle] remove cache warmers when Twig cache is disabled

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 2.8
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #28009
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php [hunk 1]
```diff
@@ -132,6 +132,11 @@ public function load(array $configs, ContainerBuilder $container)
 
         $container->getDefinition('twig')->replaceArgument(1, $config);
 
+        if (false === $config['cache']) {
+            $container->removeDefinition('twig.cache_warmer');
+            $container->removeDefinition('twig.template_cache_warmer');
```
Comment (xabbuh) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Does someone remember why we have two cache warmers? Both classes aim to solve the same problem. Is it still necessary to keep both?

**Comment by yceruto** [comment]
This part should be updated as well:
https://github.com/symfony/symfony/blob/e7f50768f107fa951f4c80dc447b893502cf2e89/src/Symfony/Bundle/TwigBundle/DependencyInjection/Compiler/ExtensionPass.php#L52-L55

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php (hunk 1 above)
Comment (weaverryan) [reply src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: I looked at this a LONG time ago for #24608, but I can't remember what my conclusion was. Iirc, they DO indeed do the same thing, but maybe one has one slightly additional feature. Not too helpful, I know :)

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/Compiler/ExtensionPass.php [hunk 2]
```diff
@@ -49,9 +49,12 @@ public function process(ContainerBuilder $container)
...
-            $paths = $container->getDefinition('twig.cache_warmer')->getArgument(2);
-            $paths[$coreThemePath] = null;
-            $container->getDefinition('twig.cache_warmer')->replaceArgument(2, $paths);
+            if ($container->hasDefinition('twig.cache_warmer')) {
+                $paths = $container->getDefinition('twig.cache_warmer')->getArgument(2);
+                $paths[$coreThemePath] = null;
+                $container->getDefinition('twig.cache_warmer')->replaceArgument(2, $paths);
+            }
+
             $container->getDefinition('twig.template_iterator')->replaceArgument(2, $paths);
```
Comment (yceruto) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/Compiler/ExtensionPass.php]: variable `$paths` is undefined if `twig.cache_warmer` doesn't exist, it must be included in the new `if` condition.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php (hunk 1 above)
Comment (yceruto) [reply src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: IIRC `TemplateCacheCacheWarmer` is the dedicated warmer for `templating` layer (if enabled), and it's necessary while this layer exists. In the other hand, `TemplateCacheWarmer` is the new version without `templating` layer/dependecy.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/Compiler/ExtensionPass.php [hunk 3]
```diff
@@ -49,10 +49,12 @@ public function process(ContainerBuilder $container)
...
 
-            $paths = $container->getDefinition('twig.cache_warmer')->getArgument(2);
-            $paths[$coreThemePath] = null;
-            $container->getDefinition('twig.cache_warmer')->replaceArgument(2, $paths);
-            $container->getDefinition('twig.template_iterator')->replaceArgument(2, $paths);
+            if ($container->hasDefinition('twig.cache_warmer')) {
+                $paths = $container->getDefinition('twig.cache_warmer')->getArgument(2);
+                $paths[$coreThemePath] = null;
+                $container->getDefinition('twig.cache_warmer')->replaceArgument(2, $paths);
+                $container->getDefinition('twig.template_iterator')->replaceArgument(2, $paths);
```
Comment (stof) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/Compiler/ExtensionPass.php]: either this service should be also removed when cache is disabled, or its configuration here should not be disabled when disabling the cache warming.

**Comment by fabpot** [comment]
@xabbuh Can you finish this PR?

**Comment by xabbuh** [comment]
comments should be addressed now

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Pull request merged**